third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements the "definitely available copies" analysis, also called "available copies" analysis (in short).
+//! This analysis is a prerequisite for the copy propagation transformation.
+//!
+//! A copy is of the form `a := b` (i.e., `a` is assigned `b`), where `a` and `b` are locals/temporaries.
+//!
+//! A definitely available copy at a given program point `P` is a copy `a := b` that has reached `P`
+//! along all possible program paths such that neither `a` nor `b` is overwritten along any of these paths.
+//! That is, `a` and `b` are always available unmodified at `P` after the copy `a := b`,
+//! making it definitely available.
+//!
+//! This is a forward "must" analysis.
+//! In a forward analysis, we reason about facts at a program point `P` using facts at its predecessors.
+//! The "must" qualifier means that the analysis facts must be true at `P`,
+//! irrespective of what path lead to `P`.
+
+use itertools::Itertools;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AbortAction, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Collection of definitely available copies.
+/// For a copy `a := b`, we store the key-value pair `(a, b)` in the internal map.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AvailCopies(BTreeMap</*dst*/ TempIndex, /*src*/ TempIndex>);
+
+impl AvailCopies {
+    /// Create a new (empty) collection of definitely available copies.
+    fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+
+    /// Make a copy `dst := src` available.
+    /// If either `dst` or `src` is in `borrowed_locals`, then the copy is not made available.
+    /// To call this method, `dst := x` should not already be available for any `x`.
+    fn make_copy_available(
```

Comment:
I think that having these two separated out is nice from an abstraction stand point (and the pre-conditions of the function are documented and checked) and future code development stand point (if we decide to do more involved things with the borrowed variables). I am keeping these abstractions for now, instead of combining them.